Fix/inventories

Faithcaio opened the pull request:
**SpongeAPI** | [SpongeCommon](https://github.com/SpongePowered/SpongeCommon/pull/1946) | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/2232)

Overview:
- `Container#getViewed()` added
- `Container#setCursor(ItemStack)` added
- `Container#getCursor:Optional<ItemStack>` added
- `Inventory#slots():List<Slot>` changed to return Slots
- `Inventory#children():List<Inventory>` instead of implementing Iterable
- `OrderedInventory` removed and its methods moved to `Inventory`
- `Interactable` -> `ViewableInventory`
- `Inventory#query(QueryOperations<?>...):Inventory` no ducktyping
- `Inventory#query(Class<T extends Inventory>):Optional<T>` returns none when not exact match
- `Inventory#asViewable():Optional<ViewableInventory>` 
- `Slot#transform` -> `Slot#viewedSlot` 
- `MainPlayerInventory` is not a Grid anymore `MainPlayerInventory#asGrid():GridInventory`
- `ViewableInventory#getViewers():Set<Players>` moved from `Container`
- `ViewableInventory#hasViewers():boolean` moved from `Container`
- `InventoryTransactionResult` removed unused UNDEFINED and CANCELLED Types
- `InventoryTransactionResult#getSlotTransactions` replaces `replaced` item list.
- `InventoryTransactionResult#revert()` and `InventoryTransactionResult#revertOnFailure()`
- `InventoryTransactionResult` and Builder moved to Impl
- `Inventory#peek` `Inventory#poll` do not return Optional anymore `ItemStack#empty()` instead
    - Variants with SlotIndex still return `Optional#empty` when the Slot does not exist.
- Updated/Simplified javadocs on `peek` `poll` `offer` `set`
- `UserInventory#getEquipment` now also has both main and offhand
- `UserInventory#getArmor()`
- `PrimaryPlayerInventory` renamed from MainPlayerInventory
- `UserInventory#getPrimary:PrimaryPlayerInventory` (previously `getMain`)
- `UserInventory#getStorage:GridInventory` (previously `getMainGrid`)
- `PrimaryPlayerInventory#getStorage:GridInventory` (previously `getMain`)

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -72,40 +75,18 @@
+     * @return a list of all direct child inventories.
      */
-    <T extends Inventory> T next();
+    List<Inventory> subInventories();
```
`children`?

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/type/Interactable.java:
```diff
@@ -26,12 +26,45 @@
+     * @param viewer The viewer to show this inventory to
+     * @throws IllegalArgumentException if a {@link PluginContainer} is not the root of the cause
+     */
+    void open(Player viewer) throws IllegalArgumentException;
```
I would get rid of these and force devs to use `Player#openInventory` and `Player#closeInventory` instead.

Faithcaio replied:
it really just redirects to exactly that

Cybermaxke replied:
Then there is no need to keep them?

NeumimTo commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -223,6 +224,63 @@
+     * @param stack Stack to insert
+     * @return matching stacks, as per the semantics of {@link Inventory#set}
+     */
+    InventoryTransactionResult set(SlotIndex index, ItemStack stack);
```
Would it be possible to also introduce method 

InventoryTransactionResult set(SlotPos slotPos, ItemStack stack);

as well?

Cybermaxke replied:
`SlotPos` is only used in `Inventory2D` (grid, row or column).

Faithcaio replied:
SlotPos is only available on Grids. Not all Inventories are Grids. 
So not possible

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/type/Interactable.java:
```diff
@@ -26,12 +26,45 @@
+     *
+     * @return The current viewers of this inventory
+     */
+    Set<Player> getViewers();
```
Maybe rename to `getInteractors` to match the class name, can also avoid confusion without the difference between "viewing" and being able to modify the inventory. The same for `hasViewers`.

Faithcaio replied:
This is basicly a List of players having a Container open for this Inventory.
Maybe instead rename this interface to ViewableInventory?

Cybermaxke replied:
That sounds good.

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/type/Interactable.java:
```diff
@@ -26,12 +26,45 @@
 /**
  * Interface for inventories which may be interacted with by Players.
  */
 public interface Interactable extends Inventory {
```
Rename to `InteractableInventory`? To avoid possible conflicts with naming.

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -498,6 +534,15 @@ default Inventory transform(InventoryTransformation transformation) {
+     * @return This inventory as an interactable inventory if possible.
+     */
+    Optional<Interactable> asInteractable();
+
```
Maybe in case of a large grid inventory you can also provide a method like `asInteractableView` which takes the min and max coordinates to create view for.
E.g. for a 50x50 inventory, you can get a 3x9 view starting at row 4 and column 9, and display that to the player
Just an idea though

Faithcaio replied:
I'd rather not add features with this. I'll have enough work to fix impl.
But I'll keep this in mind for later

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/Slot.java:
```diff
@@ -50,20 +38,6 @@
-    Slot transform(Type type);
-
     /**
      * Transforms this Slot into the default Type.
```
Javadoc will need to be updated.

XakepSDK commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -243,6 +224,63 @@
+     * @param index slot index to query
+     * @return matching stacks, as per the semantics of {@link Inventory#poll()}
+     */
+    Optional<ItemStack> poll(SlotIndex index);
```
Maybe make these methods default?
```java
default Optional<ItemStack> poll(SlotIndex index) {
    return query(index.asQuery()).peek();
}
```

Faithcaio replied:
no its not the same

XakepSDK commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -243,6 +224,63 @@
+     *
+     * @see Inventory#poll()
+     * @param index slot index to query
+     * @param limit item limit
```
What means this limit?

Faithcaio replied:
see `Inventory#poll`
```
Get and remove up to <code>limit</code> items of the type in the first
available stack in this Inventory from all stacks in this Inventory...
```

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -421,12 +459,21 @@
+     * @param <T> The Type of inventory
+     * @return the query result
+     */
+    <T extends Inventory> Optional<T> query(Class<T> inventoryType);
```
Maybe rename this to `querySingle` (or something else)? The normal `query` method returns a collection `Inventory` and this method returning a single `Inventory` could make the naming confusing.

Cybermaxke replied:
@Faithcaio thoughts?

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/transaction/InventoryTransactionResult.java:
```diff
@@ -74,50 +76,84 @@ public static InventoryTransactionResult failNoTransactions() {
+
+    final Type type;
 
     InventoryTransactionResult(Builder builder) {
```
Maybe move the `InventoryTransactionResult` impl to SpongeCommon?

parlough replied:
Yes please ^

gabizou commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -243,6 +224,63 @@
     InventoryTransactionResult set(ItemStack stack);
 
+    /**
+     * Gets and remove the stack at the supplied index in this Inventory.
```
These methods need to be clear on when an `Optional` is empty or not.

Because, we can simplify this to `ItemStack.EMPTY` if we're returning it for a valid slot index, and `Optional.empty()` when the `SlotIndex` is invalid. OR, simplify these methods to always return an `ItemStack` or throw an exception if the slot index is invalid. My two cents on the matter.
The optionals are somewhat annoying to work with, and frankly, it's easier to say "Hey, you're working on an empty itemstack," versus "Hey, you might be working with an empty itemstack anyways, but you have to check the optional first!"

Faithcaio replied:
I changed it to `ItemStack` instead of `Optional` for the ones without SlotIndex.
With SlotIndex still returns Optional because the Slot might not even exist.
`inv.peek(SlotIndex.of(1))` is the same as `inv.getSlot(SlotIndex.of(1)).map(Slot::peek)`

Cybermaxke commented:
I was discussing with @parlough about the fact that a `EquipmentInventory` which is located in a `PlayerInventory` doesn't contain the held item slots, in Mumfreys original PR there was spoken about a armor inventory (which are the 4 slots) and the `EquipmentInventory` should contain all the slots (ofc only the supported one).
Knowing that we could expose a `ArmorInventory` in the `PlayerInventory`, and make the `EquipmentInventory` return a inventory view containing the held and armor slots.

```java
interface PlayerInventory {

    ArmorInventory getArmor();

    EquipmentInventory getEquipment();

    // ...
}
```

Zidane commented on src/main/java/org/spongepowered/api/item/inventory/Container.java:
```diff
@@ -24,60 +24,50 @@
+     * @return The players cursor item.
      */
-    boolean isViewedSlot(Slot slot);
+    Optional<ItemStack> getCursor();
```
Should we consider empty stacks for this?

Faithcaio replied:
Its actually Optional  for the case you have a Container which has been closed already.
If no player is looking at it there is no cursor.

Zidane commented on src/main/java/org/spongepowered/api/item/inventory/Container.java:
```diff
@@ -24,60 +24,50 @@
+     * @param item The item to set.
      */
-    void close(Player viewer) throws IllegalArgumentException;
+    void setCursor(ItemStack item);
```
Could we mark parameter as Nullable and just set the cursor to empty stack then?

Faithcaio replied:
We could but should we? I'd rather not have null items anymore.

Optional is for when a Slot actually does not exists. Not for empty Slots. 
See changes to `Inventory#peek()` etc.

Zidane commented on src/main/java/org/spongepowered/api/item/inventory/Container.java:
```diff
@@ -24,60 +24,50 @@
+     * @return the list of viewed inventories.
      */
-    void open(Player viewer) throws IllegalArgumentException;
+    List<Inventory> getViewed();
```
Inventories are composite by nature, why is this needed when you can just query what you want?

Zidane commented on src/main/java/org/spongepowered/api/item/inventory/entity/UserInventory.java:
```diff
@@ -59,15 +59,13 @@ default Hotbar getHotbar() {
      */
     default GridInventory getMainGrid() {
-        return this.getMain().getGrid();
+        return this.getMain().getMain();
```
Naming seems odd here.

Faithcaio replied:
I agree and am open for suggestions

Faithcaio replied:
`getPrimary().getStorage()`

Zidane commented on src/main/java/org/spongepowered/api/item/inventory/Inventory.java:
```diff
@@ -30,20 +30,22 @@
  * Base interface for queryable inventories.
  */
-public interface Inventory extends Iterable<Inventory>, Nameable, PropertyHolder {
+public interface Inventory extends Nameable, PropertyHolder {
```
I want to hear the arguments for this change. It was at the heart of the Inventory API from the get go to not impose any presumption of ordering and, instead, leave that to the OrderedInventory interface (as that is more a Container concept).

Faithcaio replied:
The reason is simple:
All Inventories that we currently have implementations for are accessible by index and getting a Slot of an Inventory by index was a lot of work.
You either had to check&cast to OrderedInventory, which only works for some inventories.
Or query for it but...
When you query for SlotIndex.of(0) you wont always get a single Slot instead all Slots that have index 0 recursivly. That includes the first Slot of every Row/Column.
When you query for OrderedInventory you wont always get one OrderedInventory instead EmptyInventory/OrderedInventory/multiple OrderedInventory.

Iterable was moved to `children()` instead. Because a lot of ppl. found it confusing.

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/entity/UserInventory.java:
```diff
@@ -42,32 +42,32 @@
      * @return The main inventory
      */
-    MainPlayerInventory getMain();
+    MainPlayerInventory getPrimary();
```
What is wrong with `getMain` here?
If you change this then should the transformation `PLAYER_MAIN_HOTBAR_FIRST` (https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/item/inventory/InventoryTransformations.java#L42) be renamed to `PLAYER_PRIMARY_HOTBAR_FIRST` to be consistent

kashike replied:
`getPrimary` is a nicer name, in my opinion.

Cybermaxke replied:
Maybe also renaming the class then? `PrimaryPlayerInventory`?

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/transaction/InventoryTransactionResult.java:
```diff
@@ -175,63 +156,41 @@ public Builder type(final Type type) {
+         * @param itemStacks The itemstacks being rejected
+         * @return This builder, for chaining
+         */
+        Builder reject(List<ItemStackSnapshot> itemStacks);
```
Maybe change the `List` to `Iterable`?

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/transaction/InventoryTransactionResult.java:
```diff
@@ -175,63 +156,41 @@ public Builder type(final Type type) {
-            }
-            return this;
-        }
+        Builder transaction(List<SlotTransaction> slotTransactions);
```
`List` to `Iterable`?

Faithcaio committed:
Big Inventory API Update

Removing Nullable ItemStacks and Empty Optional ItemStacks for empty slots. (Still exists for missing slots)
Add indexed access to base Inventory (removed OrderedInventory)
Changed some Inventory Names
Better InventoryTransactionResults
Viewed inventory and cursor is available on Container
Adjusted Inventory return types (no ducktyping and more specific if possible)
and some JavaDocs

XakepSDK commented:
ViewableInventory confuses me.
Player can view container, buy how can he view a inventory?
How can i get container viewers?

Faithcaio commented:
Stuff you can look into like Chests are/have `ViewableInventory`
When you look into a Chest you open a `Container`.
If two people look into the same Chest they each have their own Container.

I'll add a `getViewer:Optional<Player>` to Container

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/entity/PrimaryPlayerInventory.java:
```diff
@@ -40,10 +41,18 @@
+     * @return The grid of the players main inventory
+     */
+    GridInventory getStorage();
+
```
Unnecessary empty line

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/entity/PrimaryPlayerInventory.java:
```diff
@@ -40,10 +41,18 @@
+    GridInventory getStorage();
+
+
+}
```
Missing newline

Cybermaxke commented on src/main/java/org/spongepowered/api/item/inventory/entity/UserInventory.java:
```diff
@@ -32,7 +32,7 @@
+ * It consists of a {@link PrimaryPlayerInventory} (containing the main {@link GridInventory}
  * and the {@link Hotbar}) and an {@link EquipmentInventory}
  */
 public interface UserInventory<T extends User> extends CarriedInventory<T> {
```
Thinking about the armor and equipment inventory, what would be returned when querying for a `EquipmentInventory` class?

Faithcaio replied:
It should detect that all armor slots are contained in the other. 
So you should get one EquipmentInventory (the bigger one) back

XakepSDK commented:
Bump: no method `offer(SlotIndex, ItemStack)`

ryantheleach commented:
@XakepSDK commenting on merged PR's has a high chance of being missed. Create a new issue, contact faith to see what they think (since this was their PR), or comment on https://github.com/SpongePowered/SpongeAPI/issues/1000 for eventual cleanup / addition.
